cloud-aws/src/main/java/com/sequenceiq/cloudbreak/cloud/aws/AwsIDBrokerObjectStorageValidator.java

```diff
@@ -0,0 +1,258 @@
+import com.amazonaws.services.identitymanagement.model.SimulatePrincipalPolicyRequest;
+import com.amazonaws.services.identitymanagement.model.SimulatePrincipalPolicyResult;
+import com.sequenceiq.cloudbreak.cloud.exception.CloudConnectorException;
+import com.sequenceiq.cloudbreak.cloud.model.filesystem.CloudFileSystemView;
+import com.sequenceiq.cloudbreak.cloud.model.filesystem.CloudS3View;
+import com.sequenceiq.common.api.cloudstorage.AccountMappingBase;
+import com.sequenceiq.common.api.cloudstorage.StorageLocationBase;
+import com.sequenceiq.common.model.CloudIdentityType;
+import com.sequenceiq.common.model.CloudStorageCdpService;
+
+public class AwsIDBrokerObjectStorageValidator {
+    private static final Logger LOGGER = LoggerFactory.getLogger(AwsIDBrokerObjectStorageValidator.class);
+
+    private final AmazonIdentityManagement iam;
+
+    public AwsIDBrokerObjectStorageValidator(AmazonIdentityManagement amazonIdentityManagement) {
+        this.iam = amazonIdentityManagement;
+    }
+
+    public void validateObjectStorage(List<CloudFileSystemView> cloudFileSystems) {
+        for (CloudFileSystemView cloudFileSystemView : cloudFileSystems) {
+            CloudS3View cloudFileSystem = (CloudS3View) cloudFileSystemView;
+            InstanceProfile instanceProfile = getInstanceProfile(cloudFileSystem.getInstanceProfile());
+            if (instanceProfile != null) {
+                CloudIdentityType cloudIdentityType = cloudFileSystem.getCloudIdentityType();
+                if (CloudIdentityType.ID_BROKER.equals(cloudIdentityType)) {
+                    checkInstanceProfileTrust(instanceProfile);
+                    checkIDBrokerAssumeRolePermissions(instanceProfile, cloudFileSystem);
+                    checkDataAccessRolePermissions(cloudFileSystem);
+                    checkRangerAuditRolePermissions(cloudFileSystem);
+                } else if (CloudIdentityType.LOG.equals(cloudIdentityType)) {
+                    checkInstanceProfileTrust(instanceProfile);
+                }
+            }
+        }
+    }
+
+    private InstanceProfile getInstanceProfile(String instanceProfileArn) {
+        InstanceProfile instanceProfile = null;
+        if (instanceProfileArn != null && instanceProfileArn.contains("/")) {
+            String instanceProfileName = instanceProfileArn.split("/", 2)[1];
+            GetInstanceProfileRequest instanceProfileRequest = new GetInstanceProfileRequest().withInstanceProfileName(instanceProfileName);
+            try {
+                instanceProfile = iam.getInstanceProfile(instanceProfileRequest).getInstanceProfile();
+            } catch (NoSuchEntityException | ServiceFailureException e) {
+                throw new CloudConnectorException(String.format("Unable to find instance profile: %s",
+                    instanceProfileArn), e);
+            }
+        }
+        return instanceProfile;
+    }
+
+    private void checkDataAccessRolePermissions(CloudS3View cloudFileSystem) {
```

Comment:
Each of these private methods are pretty large and do relatively complex checks. Extracting them to separate classes of their own, and then unit testing each of them would be great. Suggestion with referring to my previous comment: The input and output of the methods of these validator classes could be the `ValidationResultBuilder` in which you could collect the validation results.